src/meta-srv/src/service/heartbeat.rs

```diff
@@ -48,10 +51,18 @@ impl heartbeat_server::Heartbeat for MetaSrv {
                 let mut quit = false;
                 match msg {
                     Ok(req) => {
-                        let role = req.header.as_ref().map_or(0, |h| h.role);
+                        let header = match req.header.as_ref() {
+                            Some(header) => header,
+                            None => {
+                                warn!("Refused to process the heartbeat request, because the header is missing.");
+                                continue;
+                            }
+                        };
+
                         if pusher_key.is_none() {
-                            if let Some(peer) = &req.peer {
-                                let key = format!("{}-{}", role, peer.id,);
+                            if let Some(node_id) = get_node_id(&req) {
```

Comment:
I think `get_node_id` should not return `None`. Every Datanode must carries a node id, otherwise it's an error.